google_guest_agent/network/manager/network_manager_linux.go

```diff
@@ -0,0 +1,268 @@
+	// GuestAgent is the 'guest-agent' section.
+	GuestAgent nmGuestAgentSection `ini:"guest-agent"`
+
+	// Connection is the connection section.
+	Connection nmConnectionSection `ini:"connection"`
+
+	// Ipv4 is the ipv4 section.
+	Ipv4 nmIpv4Section `ini:"ipv4"`
+
+	// Ipv6 is the ipv6 section.
+	Ipv6 nmIpv6Section `ini:"ipv6"`
+}
+
+// networkManager implements the manager.Service interface for NetworkManager.
+type networkManager struct {
+	// configDir is the directory to which to write the configuration files.
+	configDir string
+}
+
+// init registers this network manager service to the list of known network managers.
+func init() {
+	registerManager(&networkManager{
+		configDir: defaultNetworkManagerConfigDir,
+	}, false)
+}
+
+// Name is the name of this network manager service.
+func (n networkManager) Name() string {
+	return "NetworkManager"
+}
+
+// IsManaging checks if NetworkManager is managing the provided interface.
+func (n networkManager) IsManaging(ctx context.Context, iface string) (bool, error) {
+	// Check the status of network.service. This mainly works on SUSE distributions.
+	res := run.WithOutput(ctx, "systemctl", "status", "network.service")
+	if strings.Contains(res.StdOut, "NetworkManager.service") {
+		return true, nil
+	}
+
+	// Check for existence of nmcli. Without nmcli, the agent cannot tell NetworkManager
+	// to reload the configs for its connections.
+	if _, err := execLookPath("nmcli"); err != nil {
+		if errors.Is(err, exec.ErrNotFound) {
+			return false, nil
+		}
+		return false, fmt.Errorf("error checking for nmcli: %v", err)
+	}
+
+	// Check whether NetworkManager.service is active.
+	if err := run.Quiet(ctx, "systemctl", "is-active", "NetworkManager.service"); err != nil {
+		return false, nil
+	}
+
+	// Use nmcli to check status of provided  interface.
+	res = run.WithOutput(ctx, "nmcli", "-f", "DEVICE,STATE", "dev", "status")
+	if res.ExitCode != 0 {
+		return false, fmt.Errorf("error checking status of devices on NetworkManager: %v", res.StdErr)
+	}
+
+	output := res.StdOut
+	lines := strings.Split(output, "\n")
+	for _, line := range lines {
+		if strings.HasPrefix(line, iface) {
+			fields := strings.Fields(line)
+			return fields[1] == "connected", nil
+		}
+	}
+	return false, nil
+}
+
+// Setup sets up the necessary configurations for NetworkManager.
+func (n networkManager) Setup(ctx context.Context, config *cfg.Sections, payload []metadata.NetworkInterfaces) error {
+	ifaces, err := interfaceNames(payload)
+	if err != nil {
+		return fmt.Errorf("error getting interfaces: %v", err)
+	}
+
+	connections, err := n.writeNetworkManagerConfigs(ifaces)
+	if err != nil {
+		return fmt.Errorf("error writing NetworkManager connection configs: %v", err)
+	}
+
+	// Load all the connections.
+	for _, iface := range ifaces {
+		configFile := n.networkManagerConfigFilePath(iface)
+
+		// NetworkManager can only load and use config files with 600 permission.
+		if err := os.Chmod(configFile, 0600); err != nil {
```

Comment:
I'd rather have this `Chmod()` call closer to where the file is created i.e. `writeNetworkManagerConfigs()`.

Reply:
Done.